### ISSUE TITLE
Embed metadata into the wasm binary

### ISSUE THREAD

**Issue by pepyakin** [paritytech/polkadot-sdk#2745]
> [!NOTE]
> This is a copy of https://github.com/paritytech/substrate/issues/8870 that was not copied over, but I still believe is worthwhile to persue.

https://github.com/paritytech/substrate/issues/7327 and https://github.com/paritytech/substrate/pull/8688 were about embedding the version information into the wasm binary. We can leverage the same technique for the metadata.

That would enable a simpler way to get the metadata. The user will need to only obtain the wasm binary and then fetch a specific section. No instantiation of the wasm runtime is required.

**Comment by xlc** [comment]
This may or maynot be blocked by https://github.com/paritytech/polkadot-sdk/issues/1139

**Comment by pepyakin** [comment]
Oh god, I am definitely on the side that the metadata should be constant.

(also see the linked issues for the motivation)

**Comment by bkchr** [comment]
Can be done as part of https://github.com/paritytech/polkadot-sdk/issues/291, because there we will already need to fetch the metadata from the built runtime.

**Comment by pepyakin** [comment]
Agree! That said, I would still leave this one opened until addressed.
